### PR TITLE
Fix race condition causing 'pool not open' errors

### PR DESCRIPTION
**Pull request by gsheasby**
**Goals (and why)**:
In PDS-239577, we had a report of various failures, and the 'pool not open' error was a common theme. We know of a race condition where, on removing a host from the client pool, we shut down the pooling container _before_ removing the host from the map. This means that there exists a short time when hosts can be present in the pool, but closed (and thus unable to serve requests).

==COMMIT_MSG==
Fixed a race condition where we would potentially attempt to use a Cassandra client pool that we had already closed.
==COMMIT_MSG==

**Implementation Description (bullets)**:
- Fix race condition
- Drive-by improvement to logging

**Testing (What was existing testing like?  What have you done to improve it?)**: Tricky to test for this; we would expect to see a reduction in failure frequency once this has rolled out.

**Concerns (what feedback would you like?)**: Is there more we can do to avoid thrashing of client pools?

**Where should we start reviewing?**: +14/-5

**Priority (whenever / two weeks / yesterday)**: yesterday

### CODE REVIEW

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java [hunk 1]
```diff
@@ -399,15 +399,16 @@ public void addPool(InetSocketAddress server) {
 
     public void removePool(InetSocketAddress removedServerAddress) {
         blacklist.remove(removedServerAddress);
+        CassandraClientPoolingContainer containerToRemove = currentPools.get(removedServerAddress);
+        currentPools.remove(removedServerAddress);
         try {
-            currentPools.get(removedServerAddress).shutdownPooling();
+            containerToRemove.shutdownPooling();
```
Comment (Jolyon-S) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java:405]: technically doesn't fully solve the issue: what if there are in-flight requests on the pool-to-be-closed?

Comment (gsheasby) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java:405]: hmm, yes - ideally we'd have something that will:
a) block new requests from being started
b) wait for existing requests to complete (or time out)
c) clean up by shutting down pooling.

Comment (gsheasby) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java:405]: Looks like once we've called `borrowObject` (in `CassandraClientPoolingContainer.runWithGoodResource`), we should be good?

I'm not averse to adding something like this if it's proven to be necessary (and I can follow up this PR with logging to help us to determine this), but for now I think in-flight requests should continue to work.

Comment (Jolyon-S) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java:405]: That's pretty nice, actually!

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java [hunk 2]
```diff
@@ -399,15 +399,16 @@ public void addPool(InetSocketAddress server) {
 
     public void removePool(InetSocketAddress removedServerAddress) {
         blacklist.remove(removedServerAddress);
+        CassandraClientPoolingContainer containerToRemove = currentPools.get(removedServerAddress);
+        currentPools.remove(removedServerAddress);
```
Comment (Jolyon-S) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/pool/CassandraService.java:403]: `remove` returns the value removed, so we shouldn't need the get as well.

**Review by Jolyon-S — APPROVED** [review]
LGTM! Just one nit, but otherwise nice! Also, your PR title has a typo

**Pull request merged**